emcc.py

```diff
@@ -2753,19 +2759,43 @@ def do_binaryen(target, asm_target, options, memfile, wasm_binary_target,
   if shared.Settings.SIDE_MODULE:
     sys.exit(0) # and we are done.
 
-  if options.opt_level >= 2:
+  if options.opt_level >= 2 and options.debug_level <= 2:
     # minify the JS
     optimizer.do_minify() # calculate how to minify
-    if optimizer.cleanup_shell or options.use_closure_compiler:
-      save_intermediate_with_wasm('preclean', wasm_binary_target)
-      final = shared.Building.minify_wasm_js(js_file=final,
-                                             wasm_file=wasm_binary_target,
-                                             expensive_optimizations=will_metadce(options),
-                                             minify_whitespace=optimizer.minify_whitespace,
-                                             use_closure_compiler=options.use_closure_compiler,
-                                             debug_info=debug_info,
-                                             emit_symbol_map=emit_symbol_map)
-      save_intermediate_with_wasm('postclean', wasm_binary_target)
+    save_intermediate_with_wasm('preclean', wasm_binary_target)
+    final = shared.Building.minify_wasm_js(js_file=final,
+                                           wasm_file=wasm_binary_target,
+                                           expensive_optimizations=will_metadce(options),
+                                           minify_whitespace=optimizer.minify_whitespace,
+                                           debug_info=debug_info,
+                                           emit_symbol_map=emit_symbol_map)
+    save_intermediate_with_wasm('postclean', wasm_binary_target)
+
+  def run_closure_compiler(final):
+    final = shared.Building.closure_compiler(final, pretty=not optimizer.minify_whitespace)
+    save_intermediate_with_wasm('closure', wasm_binary_target)
+    return final
+
+  # In the general case, run closure here. wasm2js is more complicated: in mode 1,
+  # we run closure here, so that it does not affect the wasm2js output, which is
+  # added later. In mode 2, we do run it afterwards, so it sees the entire program
+  # including the wasm2js output.
+  # TODO: is it worth it?
+  if options.use_closure_compiler and (not shared.Settings.WASM2JS or options.use_closure_compiler == 1):
+    final = run_closure_compiler(final)
+
+  if shared.Settings.WASM2JS:
+    final = shared.Building.wasm2js(final,
+                                    wasm_binary_target,
+                                    opt_level=options.opt_level,
+                                    minify_whitespace=optimizer.minify_whitespace,
+                                    use_closure_compiler=options.use_closure_compiler,
+                                    debug_info=debug_info)
+    save_intermediate_with_wasm('wasm2js', wasm_binary_target)
+
+    if options.use_closure_compiler and options.use_closure_compiler == 2:
```

Comment:
just `if options.use_closure_compiler == 2`? The first one is redundant

Also, does closure mode 2 imply that we should run this twice? Because that's what it seems to be doing.

Oh, maybe this should be `if WASM2JS and closure == 2`

Reply:
Thanks, yeah, this can be just `closure == 2` (since we are inside a wasm2js block here), I'll fix.

The idea with closure mode 2 is that we run closure on all the code together - compiled and non-compiled. Otherwise we minify the two separately. (It's not actually clear closure 2 helps much here, though, I'm still investigating.) This should be the single spot we run closure, when we are in mode 2 (the call earlier up has a check to not run it there as well).